### DISCUSSION TITLE
How to use Gnome Passwords / Keyring with Forge

### DISCUSSION THREAD

**Discussion by zarthross** [magit/forge#462] [Show and tell]
Posting this here because it might help someone else, or might make its way into the manual:

Add this to your init to instruct emacs to use the secrets api (aka gnome passwords/keyring)
```
;;; use the Secrets API Login collection
;;; (see Secret Service API)
(setq auth-sources '("secrets:Login"))
```

You need to eval this after you set the auth source as above, to create a secret in the gnome keyring with the fields you want since you cannot do this from the UI: 
```
;;; 
(secrets-create-item "login" "Forge Github Token" "YOUR_GITHUB_TOKEN_GOES_HERE"
 :host "api.github.com"
 :user "YOUR_USERNAME_GOES_HERE^forge"
 )
```

**Comment by tarsius** [comment]
Thanks!

I think that would make a nice addition to https://github.com/magit/forge/wiki/Tips-and-Tricks. (But use only two semicolons for comments (three and more are for outline headings) and don't put the closing parenthesis on its own line; that makes certain people nervous.)

**Reply by tarsius** [reply]
I've added it there now.

**Comment by zoechi** [comment]
I just tried this with KeepassXC and it worked on GitLab.
Then I tried it on GitHub and it failed with `401 - Unauthorized - Bad credentials`.

I then used this call from comment https://github.com/magit/forge/discussions/433#discussioncomment-1614560

```lisp
(ghub--token "api.github.com" "zoechi" 'forge)
```
and it turned out it uses the GitLab token instead the GitHub one.

I passed the `:host "api.github.com"` to `(secrets-create-item ...)`, but it doesn't seem to take it into account when requesting the token.

```lisp
(secrets-search-items "keyring" :user "zoechi^forge" :host "api.github.com")
(secrets-search-items "keyring" :user "zoechi^forge" :host "gitlab.com/api/v4")

(auth-source-search :host "api.github.com" :user "zoechi" :create t)
(auth-source-search :host "gitlab.com/api/v4" :user "zoechi" :create t)
```
All return correct entries.  
When I omit the `:host "..."` part, it returns 2 entries with GitLab's being the first (auth-source-search asks for password).

Any idea how I could solve this without changing one user name? (I'm not even sure different usernames would fix this)

---

## Edit: 

Sorry for the noise.  
It is working fine.  
The issue was caused by an outdated `~/.ssh/known_hosts` entry for github.com
